engine/src/main/java/org/camunda/bpm/engine/impl/batch/AbstractBatchJobHandler.java

```diff
@@ -80,6 +89,48 @@ public boolean createJobs(BatchEntity batch) {
     return deploymentAware ? idMappings.isEmpty() : ids.isEmpty();
   }
 
+  @Override
+  public final void execute(final BatchJobConfiguration configuration,
+                            final ExecutionEntity execution,
+                            final CommandContext commandContext,
+                            final String tenantId) {
+
+    commandContext.getDbEntityManager().registerOptimisticLockingListener(new OptimisticLockingListener() {
+      @Override
+      public Class<? extends DbEntity> getEntityType() {
+        return BatchEntity.class;
+      }
+
+      @Override
+      public OptimisticLockingResult failedOperation(final DbOperation operation) {
+        if (operation instanceof DbEntityOperation) {
+          return OptimisticLockingResult.IGNORE;
+        }
+        return OptimisticLockingResult.THROW;
+      }
+    });
+
+    final String jobDefinitionId = commandContext.getCurrentJob().getJobDefinitionId();
+    final BatchEntity batch = commandContext.getBatchManager().findBatchByJobDefinitionId(jobDefinitionId);
+    if (batch != null && batch.getExecutionStartTime() == null) {
+      final HistoricBatchEntity historicBatch = commandContext.getHistoricBatchManager().findHistoricBatchById(batch.getId());
```

Comment:
Refactored it. I noticed that not all of the json converters had set the `batchId` so I introduced a new abstract object converter (`AbstractBatchConfigurationObjectConverter`) to always set the batchId for all the converters.
I also moved the deletion of the config byte array into the abstract class.

Reply:
Might adding the `batchId` now and not having it in older versions be problematic in migration scenarios?

Reply:
Well if customers have jobs in progress and they update to 7.18 then the `batchId` won't be available for a few jobs. These jobs would simply not update the `executionStartTime`.
All the jobs that are started after the 7.18 update would work fine. I think that's OK.
What do you think?
As an alternative I could add a fallback to the `findBatchByJobDefinitionId` select statement.

Reply:
I agree that the impact is so small that it is most likely neglectable. To keep our codebase lean, let's not introduce a fallback. However, let's add a paragraph to the migration guide that batches created before 7.18 won't get an execution start time assigned if created before 7.18. Another thing to consider would be a rolling update. However, since we [don't support using new features during a rolling update](https://docs.camunda.org/manual/7.17/update/rolling-update/#usage-of-new-features), that shouldn't be a problem.

Reply:
Maybe we can add another paragraph to the migration guide that batch operations are not allowed to be started during a rolling update due to the new execution start time feature. Or wouldn't it be a problem to execute batch jobs with the old engine when the new property is present in the configuration column?

Reply:
It shouldn't be a problem because the code just reads out specific fields with the `JsonUtil`. Those fields are still the same but there is an extra `batchId` which would be ignored by most of the handlers in the old version.

Reply:
An interesting case: if they update to 7.18 and have jobs that do have `batchId` in the config then after the update the `executionStartTime` will be set to the first job's execution date that runs after the update. Which might not actually be the correct execution start time since the batch was started before the update :)

Reply:
Good point. Let's document this in the migration guide as well. I don't think this is a big deal when it is transparent to users.